presto-main/src/main/java/com/facebook/presto/operator/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,126 @@
+import static com.facebook.presto.operator.UncheckedByteArrays.setLongUnchecked;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static sun.misc.Unsafe.ARRAY_LONG_INDEX_SCALE;
+
+public class LongArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    @VisibleForTesting
+    public static final int POSITION_SIZE = Long.BYTES + Byte.BYTES;
+
+    private static final String NAME = "LONG_ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(LongArrayBlockEncodingBuffers.class).instanceSize();
+
+    private byte[] valuesBuffer;
+    private int valuesBufferIndex;
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        valuesBufferIndex = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        throw new UnsupportedOperationException("accumulateRowSizes is not supported for fixed width types");
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += (positionOffsets[i + 1] - positionOffsets[i]) * POSITION_SIZE;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendValuesToBuffer();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    @Override
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        output.writeInt(bufferedPositionCount);
+
+        serializeNullsTo(output);
+
+        if (valuesBufferIndex > 0) {
+            output.appendBytes(valuesBuffer, 0, valuesBufferIndex);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPositionsRetainedSizeInBytes() +
+                sizeOf(valuesBuffer) +
+                getNullsBufferRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                getNullsBufferSerializedSizeInBytes() +   // nulls buffer
+                valuesBufferIndex;              // values buffer
+    }
+
+    private void appendValuesToBuffer()
+    {
+        valuesBuffer = ensureCapacity(valuesBuffer, valuesBufferIndex + batchSize * ARRAY_LONG_INDEX_SCALE, 2.0f, true);
+
+        int[] positions = getPositions();
+        if (decodedBlock.mayHaveNull()) {
+            for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
+                int position = positions[i];
+                int newIndex = setLongUnchecked(valuesBuffer, valuesBufferIndex, decodedBlock.getLong(position));
```

Comment:
Curious: Another approach would be simple do 
```
newIndex += SIZE_OF_LONG;
```

Curious what's the benefit of let `setLongUnchecked` return the new index? -- Or this is the pattern in other buffer related methods?

Reply:
The original way was to use 
`newIndex += ARRAY_LONG_INDEX_SCALE; `
but I think it's better to hide the internal details accessing raw arrays from the caller, ie. caller (in this case the buffer object) doesn't need to know about how to advance the index. Performance wise, these two ways are the same. I verified it in original JMH benchmarks but removed it and only kept the ones with returned index. If you like I can added them back to the JMH benchmark.

Reply:
@yingsu00 : It's not immediately obvious to me that `setXXX` would return the position after setting the value. But I don't have strong opinion on this.